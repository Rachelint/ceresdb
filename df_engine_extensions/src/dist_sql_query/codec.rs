// Copyright 2023 The CeresDB Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::sync::Arc;

use ceresdbproto::remote_engine::{extension_node::TypedExtension, DistSqlQueryExtensionNode};
use datafusion::{
    error::{DataFusionError, Result as DfResult},
    execution::FunctionRegistry,
    physical_plan::ExecutionPlan,
};

use crate::{
    codec::TypedPhysicalExtensionCodec, dist_sql_query::physical_plan::UnresolvedSubTableScan,
};

#[derive(Debug)]
pub struct DistSqlQueryCodec;

impl TypedPhysicalExtensionCodec for DistSqlQueryCodec {
    // It is possible to have more extension physical plans, so `if let` is
    // necessary.
    #[allow(irrefutable_let_patterns)]
    fn try_decode(
        &self,
        typed_extension: &TypedExtension,
        _inputs: &[Arc<dyn ExecutionPlan>],
        _registry: &dyn FunctionRegistry,
    ) -> Option<DfResult<Arc<dyn ExecutionPlan>>> {
        if let TypedExtension::DistSqlQuery(extension) = typed_extension {
            match extension.unresolved_sub_scan.clone() {
                Some(plan_pb) => {
                    Some(UnresolvedSubTableScan::try_from(plan_pb).map(|plan| Arc::new(plan) as _))
                }
                None => Some(Err(DataFusionError::Internal(
                    "actual node not found in dist query extension plan".to_string(),
                ))),
            }
        } else {
            None
        }
    }

    fn try_encode(&self, node: Arc<dyn ExecutionPlan>) -> Option<DfResult<TypedExtension>> {
        let plan_pb_res: DfResult<DistSqlQueryExtensionNode> =
            if let Some(plan) = node.as_any().downcast_ref::<UnresolvedSubTableScan>() {
                plan.clone().try_into().map(|pb| DistSqlQueryExtensionNode {
                    unresolved_sub_scan: Some(pb),
                })
            } else {
                return None;
            };

        Some(plan_pb_res.map(TypedExtension::DistSqlQuery))
    }
}

#[cfg(test)]
mod test {
    use datafusion::{physical_plan::displayable, prelude::SessionContext};
    use datafusion_proto::bytes::{
        physical_plan_from_bytes_with_extension_codec, physical_plan_to_bytes_with_extension_codec,
    };

    use crate::{codec::PhysicalExtensionCodecImpl, dist_sql_query::test_util::TestContext};

    #[test]
    fn test_sub_table_scan_codec() {
        let test_ctx = TestContext::default();
        let sub_table_plan = test_ctx.build_basic_sub_table_plan();
        let extension_codec = PhysicalExtensionCodecImpl::default();
        let session_ctx = SessionContext::default();

        // Encode and decode again
        let encoded_plan =
            physical_plan_to_bytes_with_extension_codec(sub_table_plan.clone(), &extension_codec)
                .unwrap();
        let re_decoded_plan = physical_plan_from_bytes_with_extension_codec(
            &encoded_plan,
            &session_ctx,
            &extension_codec,
        )
        .unwrap();

        // Compare.
        let expected = displayable(sub_table_plan.as_ref())
            .indent(true)
            .to_string();
        let re_decoded = displayable(re_decoded_plan.as_ref())
            .indent(true)
            .to_string();

        assert_eq!(expected, re_decoded);
    }
}
