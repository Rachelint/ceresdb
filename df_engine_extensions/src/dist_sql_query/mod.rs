// Copyright 2023 The CeresDB Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::{fmt, sync::Arc};

use async_trait::async_trait;
use datafusion::{
    error::Result as DfResult,
    execution::TaskContext,
    physical_plan::{ExecutionPlan, SendableRecordBatchStream},
};
use futures::future::BoxFuture;
use table_engine::{
    remote::model::TableIdentifier,
    table::{ReadRequest, TableRef},
};

pub mod codec;
pub mod physical_plan;
pub mod resolver;
#[cfg(test)]
pub mod test_util;

/// Remote datafusion physical plan executor
pub trait RemotePhysicalPlanExecutor: fmt::Debug + Send + Sync + 'static {
    fn execute(
        &self,
        table: TableIdentifier,
        task_context: &TaskContext,
        plan: Arc<dyn ExecutionPlan>,
    ) -> DfResult<BoxFuture<'static, DfResult<SendableRecordBatchStream>>>;
}

type RemotePhysicalPlanExecutorRef = Arc<dyn RemotePhysicalPlanExecutor>;

/// Executable scan's builder
///
/// It is not suitable to restrict the detailed implementation of executable
/// scan, so we define a builder here which return the general `ExecutionPlan`.
#[async_trait]
pub trait ExecutableScanBuilder: fmt::Debug + Send + Sync + 'static {
    async fn build(
        &self,
        table: TableRef,
        read_request: ReadRequest,
    ) -> DfResult<Arc<dyn ExecutionPlan>>;
}

type ExecutableScanBuilderRef = Box<dyn ExecutableScanBuilder>;
