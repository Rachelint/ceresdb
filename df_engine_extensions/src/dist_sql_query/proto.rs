// Copyright 2023 The CeresDB Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::sync::Arc;

use ceresdbproto::remote_engine::dist_sql_query_extension_node;
use datafusion::{
    error::{DataFusionError, Result as DfResult},
    execution::FunctionRegistry,
    physical_plan::ExecutionPlan,
};
use datafusion_proto::{physical_plan::PhysicalExtensionCodec, bytes::{physical_plan_to_bytes_with_extension_codec, physical_plan_from_bytes_with_extension_codec}};
use prost::{Message, bytes::Bytes};

use crate::dist_sql_query::physical_plan::UnresolvedSubTableScan;

pub struct ProtobufEncoder {
    extension_codec: DistSqlQueryCodec,
}

impl ProtobufEncoder {
    fn new() -> Self {
        Self {
            extension_codec: DistSqlQueryCodec,
        }
    }

    fn encode(&self, physical_plan: Arc<dyn ExecutionPlan>) -> DfResult<Bytes> {        
        physical_plan_to_bytes_with_extension_codec(physical_plan, &self.extension_codec)
    }

    fn decode(&self, bytes: &[u8]) -> DfResult<Arc<dyn ExecutionPlan>> {
        physical_plan_from_bytes_with_extension_codec(&bytes, ctx, &self.extension_codec)
    }
}

#[derive(Debug)]
struct DistSqlQueryCodec;

impl PhysicalExtensionCodec for DistSqlQueryCodec {
    fn try_decode(
        &self,
        buf: &[u8],
        _inputs: &[Arc<dyn ExecutionPlan>],
        _registry: &dyn FunctionRegistry,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        let dist_sql_query_plan: ceresdbproto::remote_engine::DistSqlQueryExtensionNode =
            Message::decode(buf).map_err(|e| {
                DataFusionError::Internal(format!(
                    "failed to decode dist sql query extension plan, err:{e}"
                ))
            })?;

        match dist_sql_query_plan.node {
            Some(dist_sql_query_extension_node::Node::UnreolvedSubScan(plan_pb)) => {
                let plan: UnresolvedSubTableScan = plan_pb.try_into()?;
                Ok(Arc::new(plan))
            }

            None => Err(DataFusionError::Internal(format!(
                "actual node not found in dist query extension plan"
            ))),
        }
    }

    fn try_encode(&self, node: Arc<dyn ExecutionPlan>, buf: &mut Vec<u8>) -> DfResult<()> {
        let plan_node_pb = if let Some(plan) =
            node.as_any().downcast_ref::<UnresolvedSubTableScan>()
        {
            let plan_pb: ceresdbproto::remote_engine::UnreolvedSubScan = plan.clone().try_into()?;

            plan_pb
        } else {
            return Err(DataFusionError::Internal(
                "unexpected dist sql query extension plan type, expected:[ UnresolvedSubScan ]"
                    .to_string(),
            ));
        };

        let extension_plan_pb = ceresdbproto::remote_engine::DistSqlQueryExtensionNode {
            node: Some(dist_sql_query_extension_node::Node::UnreolvedSubScan(
                plan_node_pb,
            )),
        };

        extension_plan_pb.encode(buf).map_err(|e| 
            DataFusionError::Internal(format!("failed to encode dist sql query extension plan, plan:{extension_plan_pb:?}, err:{e}")))
    }
}
