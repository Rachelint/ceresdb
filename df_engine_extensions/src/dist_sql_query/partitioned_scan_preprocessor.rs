// Copyright 2023 The CeresDB Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::sync::Arc;

use datafusion::{error::Result as DfResult, physical_plan::ExecutionPlan};
use table_engine::remote::RemoteEngineRef;

use crate::dist_sql_query::{
    partitioned_table_scan::{ResolvedPartitionedScan, UnresolvedPartitionedScan},
    sub_table_scan::UnresolvedSubTableScan,
};

struct PartitionedTablePreprocessor {
    remote_engine: RemoteEngineRef,
}

impl PartitionedTablePreprocessor {
    fn resolve_plan(&self, plan: Arc<dyn ExecutionPlan>) -> DfResult<Arc<dyn ExecutionPlan>> {
        // Leave node, let's resolve it and return.
        if let Some(unresolved) = plan.as_any().downcast_ref::<UnresolvedPartitionedScan>() {
            let sub_tables = unresolved.sub_tables.clone();
            let remote_plans = sub_tables
                .into_iter()
                .map(|table| {
                    let plan = Arc::new(UnresolvedSubTableScan {
                        table: table.clone(),
                        read_request: unresolved.read_request.clone(),
                    });
                    (table, plan as _)
                })
                .collect::<Vec<_>>();

            return Ok(Arc::new(ResolvedPartitionedScan {
                remote_engine: self.remote_engine.clone(),
                remote_exec_plans: remote_plans,
            }));
        }

        let children = plan.children().clone();
        // Occur some node isn't table scan but without children? It should return, too.
        if children.is_empty() {
            return Ok(plan);
        }

        // Resolve children if exist.
        let mut new_children = Vec::with_capacity(children.len());
        for child in children {
            let child = self.resolve_plan(child)?;

            new_children.push(child);
        }

        // There may be `ResolvedPartitionedScan` node in children, try to extend such
        // children.
        Self::maybe_extend_partitioned_scan(new_children, plan)
    }

    fn maybe_extend_partitioned_scan(
        new_children: Vec<Arc<dyn ExecutionPlan>>,
        current_node: Arc<dyn ExecutionPlan>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        if new_children.is_empty() {
            return Ok(current_node);
        }

        current_node.with_new_children(new_children)
    }
}

#[cfg(test)]
mod test {
    use std::sync::Arc;

    use common_types::{projected_schema::ProjectedSchema, tests::build_schema_for_cpu};
    use datafusion::{
        logical_expr::{expr_fn, Literal, Operator},
        physical_plan::{
            displayable,
            expressions::{binary, col, lit},
            filter::FilterExec,
            projection::ProjectionExec,
            ExecutionPlan, PhysicalExpr,
        },
        scalar::ScalarValue,
    };
    use table_engine::{
        memory::MockRemoteEngine,
        predicate::PredicateBuilder,
        remote::model::TableIdentifier,
        table::{ReadOptions, ReadRequest},
    };
    use trace_metric::MetricsCollector;

    use crate::dist_sql_query::{
        partitioned_scan_preprocessor::PartitionedTablePreprocessor,
        partitioned_table_scan::UnresolvedPartitionedScan,
    };

    #[test]
    fn test_resolve_plan() {
        let plan = build_test_physical_plan();
        let preprocessor = PartitionedTablePreprocessor {
            remote_engine: Arc::new(MockRemoteEngine),
        };
        let new_plan = displayable(preprocessor.resolve_plan(plan).unwrap().as_ref())
            .indent(true)
            .to_string();
        insta::assert_snapshot!(new_plan);
    }

    fn build_test_physical_plan() -> Arc<dyn ExecutionPlan> {
        let test_schema = build_schema_for_cpu();
        let sub_tables = vec![
            "__test_1".to_string(),
            "__test_2".to_string(),
            "__test_3".to_string(),
        ]
        .into_iter()
        .map(|table| TableIdentifier {
            catalog: "test_catalog".to_string(),
            schema: "test_schema".to_string(),
            table,
        })
        .collect::<Vec<_>>();

        // Logical exprs.
        // Projection: [time, tag1, tag2, value, field2]
        let projection = vec![1_usize, 2, 3, 4, 5];
        let projected_schema =
            ProjectedSchema::new(test_schema.clone(), Some(projection.clone())).unwrap();
        // Filter: time < 1691974518000 and tag1 == 'test_tag'
        let logical_filters = vec![(expr_fn::col("time").lt(ScalarValue::TimestampMillisecond(
            Some(1691974518000),
            None,
        )
        .lit()))
        .and(expr_fn::col("tag1").eq("test_tag".lit()))];

        // Physical exprs.
        let arrow_projected_schema = projected_schema.to_projected_arrow_schema();
        let physical_projection = vec![
            (
                col("time", &arrow_projected_schema).unwrap(),
                "time".to_string(),
            ),
            (
                col("tag1", &arrow_projected_schema).unwrap(),
                "tag1".to_string(),
            ),
            (
                col("tag2", &arrow_projected_schema).unwrap(),
                "tag2".to_string(),
            ),
            (
                col("value", &arrow_projected_schema).unwrap(),
                "value".to_string(),
            ),
            (
                col("field2", &arrow_projected_schema).unwrap(),
                "field2".to_string(),
            ),
        ];

        let physical_filter1: Arc<dyn PhysicalExpr> = binary(
            col("time", &arrow_projected_schema).unwrap(),
            Operator::Lt,
            lit(ScalarValue::TimestampMillisecond(Some(1691974518000), None)),
            &arrow_projected_schema,
        )
        .unwrap();
        let physical_filter2: Arc<dyn PhysicalExpr> = binary(
            col("tag1", &arrow_projected_schema).unwrap(),
            Operator::Eq,
            lit("test_tag"),
            &arrow_projected_schema,
        )
        .unwrap();
        let physical_filter: Arc<dyn PhysicalExpr> = binary(
            physical_filter1,
            Operator::And,
            physical_filter2,
            &arrow_projected_schema,
        )
        .unwrap();

        // Build the physical plan.
        let predicate = PredicateBuilder::default()
            .add_pushdown_exprs(&logical_filters)
            .extract_time_range(&test_schema, &logical_filters)
            .build();
        let read_request = ReadRequest {
            request_id: 42.into(),
            opts: ReadOptions::default(),
            projected_schema: projected_schema.clone(),
            predicate,
            metrics_collector: MetricsCollector::default(),
        };
        let unresolved_scan = Arc::new(UnresolvedPartitionedScan {
            sub_tables,
            read_request,
        });

        let filter: Arc<dyn ExecutionPlan> =
            Arc::new(FilterExec::try_new(physical_filter, unresolved_scan).unwrap());

        Arc::new(ProjectionExec::try_new(physical_projection, filter).unwrap())
    }
}
