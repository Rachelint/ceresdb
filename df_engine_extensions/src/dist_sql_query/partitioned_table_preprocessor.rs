// Copyright 2023 The CeresDB Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::sync::Arc;

use datafusion::physical_plan::ExecutionPlan;
use table_engine::remote::RemoteEngineRef;

use crate::dist_sql_query::{
    partitioned_table_scan::{ResolvedPartitionedScan, UnresolvedPartitionedScan},
    sub_table_scan::UnresolvedSubTableScan,
};

use datafusion::error::Result as DfResult;

struct PartitionedTablePreprocessor {
    remote_engine: RemoteEngineRef,
}

impl PartitionedTablePreprocessor {
    fn resolve_plan(&self, plan: Arc<dyn ExecutionPlan>) -> DfResult<Arc<dyn ExecutionPlan>> {
        // Leave node, let's resolve it and return.
        if let Some(unresolved) = plan.as_any().downcast_ref::<UnresolvedPartitionedScan>() {
            let sub_tables = unresolved.sub_tables.clone();
            let remote_plans = sub_tables
                .into_iter()
                .map(|table| {
                    let plan = Arc::new(UnresolvedSubTableScan {
                        table: table.clone(),
                        read_request: unresolved.read_request.clone(),
                    });
                    (table, plan as _)
                })
                .collect::<Vec<_>>();

            return Ok(Arc::new(ResolvedPartitionedScan {
                remote_engine: self.remote_engine.clone(),
                remote_exec_plans: remote_plans,
            }));
        }

        let children = plan.children().clone();
        // Occur some node isn't table scan but without children? It should return, too.                
        if children.is_empty() {
            return Ok(plan);
        }
        
        // Resolve children if exist.
        let mut new_children = Vec::with_capacity(children.len());
        for child in children {
            let child = self.resolve_plan(child)?;

            new_children.push(child);
        }

        // There may be `ResolvedPartitionedScan` node in children, try to extend such children.
        Self::maybe_extend_partitioned_scan(new_children, plan)
    }

    fn maybe_extend_partitioned_scan(
        new_children: Vec<Arc<dyn ExecutionPlan>>,
        current_node: Arc<dyn ExecutionPlan>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        if new_children.is_empty() {
            return Ok(current_node);
        }

        // All the extendable nodes have just one child.
        let children_num = current_node.children().len();
        if children_num > 1 {
            return current_node.with_new_children(new_children);
        }

        let child = new_children.first().unwrap();
        if child
            .as_any()
            .downcast_ref::<ResolvedPartitionedScan>()
            .is_some()
        {
            // TODO: judge and merge the plan node able to push down.
            return Ok(child.clone());
        }

        Ok(child.clone())
    }
}
