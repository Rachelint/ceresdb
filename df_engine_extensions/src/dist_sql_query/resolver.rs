// Copyright 2023 The CeresDB Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::sync::Arc;

use async_recursion::async_recursion;
use catalog::manager::ManagerRef as CatalogManagerRef;
use datafusion::{
    error::{DataFusionError, Result as DfResult},
    physical_plan::ExecutionPlan,
};
use table_engine::{remote::model::TableIdentifier, table::TableRef};

use crate::dist_sql_query::{
    physical_plan::{ResolvedPartitionedScan, UnresolvedPartitionedScan, UnresolvedSubTableScan},
    ExecutableScanBuilderRef, RemotePhysicalPlanExecutorRef,
};

/// Resolver which makes datafuison dist query related plan executable.
///
/// The reason we define a `Resolver` rather than `physical optimization rule`
/// is: As I see, physical optimization rule is responsible for optimizing a bad
/// plan to good one, rather than making a inexecutable plan executable.
/// So we define `Resolver` to make it, it may be somthing similar to task
/// generator responsible for generating task for executor to run based on
/// physical plan.
pub struct Resolver {
    remote_executor: RemotePhysicalPlanExecutorRef,
    catalog_manager: CatalogManagerRef,
    scan_builder: ExecutableScanBuilderRef,
}

impl Resolver {
    pub fn new(
        remote_executor: RemotePhysicalPlanExecutorRef,
        catalog_manager: CatalogManagerRef,
        scan_builder: ExecutableScanBuilderRef,
    ) -> Self {
        Self {
            remote_executor,
            catalog_manager,
            scan_builder,
        }
    }

    /// Resolve partitioned scan
    pub fn resolve_partitioned_scan(
        &self,
        plan: Arc<dyn ExecutionPlan>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        // Leave node, let's resolve it and return.
        if let Some(unresolved) = plan.as_any().downcast_ref::<UnresolvedPartitionedScan>() {
            let sub_tables = unresolved.sub_tables.clone();
            let remote_plans = sub_tables
                .into_iter()
                .map(|table| {
                    let plan = Arc::new(UnresolvedSubTableScan {
                        table: table.clone(),
                        read_request: unresolved.read_request.clone(),
                    });
                    (table, plan as _)
                })
                .collect::<Vec<_>>();

            return Ok(Arc::new(ResolvedPartitionedScan {
                remote_executor: self.remote_executor.clone(),
                remote_exec_plans: remote_plans,
            }));
        }

        let children = plan.children().clone();
        // Occur some node isn't table scan but without children? It should return, too.
        if children.is_empty() {
            return Ok(plan);
        }

        // Resolve children if exist.
        let mut new_children = Vec::with_capacity(children.len());
        for child in children {
            let child = self.resolve_partitioned_scan(child)?;

            new_children.push(child);
        }

        plan.with_new_children(new_children)
    }

    #[async_recursion]
    pub async fn resolve_sub_scan(
        &self,
        plan: Arc<dyn ExecutionPlan>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        // Leave node, let's resolve it and return.
        let build_scan_opt =
            if let Some(unresolved) = plan.as_any().downcast_ref::<UnresolvedSubTableScan>() {
                let table = self.find_table(&unresolved.table)?;
                let read_request = unresolved.read_request.clone();

                Some((table, read_request))
            } else {
                None
            };

        if let Some((table, request)) = build_scan_opt {
            return self.scan_builder.build(table, request).await;
        }

        let children = plan.children().clone();
        // Occur some node isn't table scan but without children? It should return, too.
        if children.is_empty() {
            return Ok(plan);
        }

        // Resolve children if exist.
        let mut new_children = Vec::with_capacity(children.len());
        for child in children {
            let child = self.resolve_sub_scan(child).await?;

            new_children.push(child);
        }

        plan.with_new_children(new_children)
    }

    fn find_table(&self, table_ident: &TableIdentifier) -> DfResult<TableRef> {
        let catalog = self
            .catalog_manager
            .catalog_by_name(&table_ident.catalog)
            .map_err(|e| DataFusionError::Internal(format!("failed to find catalog, err:{e}")))?
            .ok_or(DataFusionError::Internal("catalog not found".to_string()))?;

        let schema = catalog
            .schema_by_name(&table_ident.schema)
            .map_err(|e| DataFusionError::Internal(format!("failed to find schema, err:{e}")))?
            .ok_or(DataFusionError::Internal("schema not found".to_string()))?;

        schema
            .table_by_name(&table_ident.table)
            .map_err(|e| DataFusionError::Internal(format!("failed to find table, err:{e}")))?
            .ok_or(DataFusionError::Internal("table not found".to_string()))
    }
}

#[cfg(test)]
mod test {

    use datafusion::physical_plan::displayable;

    use crate::dist_sql_query::test_util::TestContext;

    #[test]
    fn test_basic_partitioned_scan() {
        let ctx = TestContext::new();
        let plan = ctx.build_basic_partitioned_table_plan();
        let resolver = ctx.resolver();
        let new_plan = displayable(resolver.resolve_partitioned_scan(plan).unwrap().as_ref())
            .indent(true)
            .to_string();
        insta::assert_snapshot!(new_plan);
    }

    #[tokio::test]
    async fn test_basic_sub_scan() {
        let ctx = TestContext::new();
        let plan = ctx.build_basic_sub_table_plan();
        let resolver = ctx.resolver();
        let new_plan = displayable(resolver.resolve_sub_scan(plan).await.unwrap().as_ref())
            .indent(true)
            .to_string();
        insta::assert_snapshot!(new_plan);
    }

    #[tokio::test]
    async fn test_unprocessed_plan() {
        let ctx = TestContext::new();
        let plan = ctx.build_unprocessed_plan();
        let resolver = ctx.resolver();

        let original_plan_display = displayable(plan.as_ref()).indent(true).to_string();

        // It should not be processed by `resolve_partitioned_scan`.
        let new_plan = resolver.resolve_partitioned_scan(plan.clone()).unwrap();

        let new_plan_display = displayable(new_plan.as_ref()).indent(true).to_string();

        assert_eq!(original_plan_display, new_plan_display);

        // It should not be processed by `resolve_sub_scan_internal`.
        let new_plan = resolver.resolve_sub_scan(plan.clone()).await.unwrap();

        let new_plan_display = displayable(new_plan.as_ref()).indent(true).to_string();

        assert_eq!(original_plan_display, new_plan_display);
    }
}
