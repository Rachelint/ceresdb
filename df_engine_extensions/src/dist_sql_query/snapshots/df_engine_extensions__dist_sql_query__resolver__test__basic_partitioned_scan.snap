---
source: df_engine_extensions/src/dist_sql_query/resolver.rs
assertion_line: 229
expression: new_plan
---
ProjectionExec: expr=[time@0 as time, tag1@1 as tag1, tag2@2 as tag2, value@3 as value, field2@4 as field2]
  FilterExec: time@0 < 1691974518000 AND tag1@1 = test_tag
    ResolvedPartitionedScan: remote_exec_plans:[(TableIdentifier { catalog: "test_catalog", schema: "test_schema", table: "__test_1" }, UnresolvedSubTableScan { table: TableIdentifier { catalog: "test_catalog", schema: "test_schema", table: "__test_1" }, read_request: ReadRequest { request_id: RequestId(42), opts: ReadOptions { batch_size: 10000, read_parallelism: 8, deadline: None }, projected_schema: ProjectedSchema { original_schema: Schema { timestamp_index: 1, tsid_index: Some(0), column_schemas: ColumnSchemas { columns: [ColumnSchema { id: 1, name: "tsid", data_type: UInt64, is_nullable: false, is_tag: false, is_dictionary: false, comment: "", escaped_name: "tsid", default_value: None }, ColumnSchema { id: 2, name: "time", data_type: Timestamp, is_nullable: false, is_tag: false, is_dictionary: false, comment: "", escaped_name: "time", default_value: None }, ColumnSchema { id: 3, name: "tag1", data_type: String, is_nullable: false, is_tag: true, is_dictionary: false, comment: "", escaped_name: "tag1", default_value: None }, ColumnSchema { id: 4, name: "tag2", data_type: String, is_nullable: false, is_tag: true, is_dictionary: false, comment: "", escaped_name: "tag2", default_value: None }, ColumnSchema { id: 5, name: "value", data_type: Int8, is_nullable: false, is_tag: false, is_dictionary: false, comment: "", escaped_name: "value", default_value: None }, ColumnSchema { id: 6, name: "field2", data_type: Float, is_nullable: false, is_tag: false, is_dictionary: false, comment: "", escaped_name: "field2", default_value: None }] }, version: 1, primary_key_indexes: [0, 1] }, projection: Some([1, 2, 3, 4, 5]) }, predicate: Predicate { exprs: [BinaryExpr(BinaryExpr { left: BinaryExpr(BinaryExpr { left: Column(Column { relation: None, name: "time" }), op: Lt, right: Literal(TimestampMillisecond(1691974518000, None)) }), op: And, right: BinaryExpr(BinaryExpr { left: Column(Column { relation: None, name: "tag1" }), op: Eq, right: Literal(Utf8("test_tag")) }) })], time_range: TimeRange { inclusive_start: Timestamp(-9223372036854775808), exclusive_end: Timestamp(1691974518000) } }, metrics_collector: MetricsCollector { name: "", metrics: Mutex { data: [], poisoned: false, .. }, children: Mutex { data: [], poisoned: false, .. } } } }), (TableIdentifier { catalog: "test_catalog", schema: "test_schema", table: "__test_2" }, UnresolvedSubTableScan { table: TableIdentifier { catalog: "test_catalog", schema: "test_schema", table: "__test_2" }, read_request: ReadRequest { request_id: RequestId(42), opts: ReadOptions { batch_size: 10000, read_parallelism: 8, deadline: None }, projected_schema: ProjectedSchema { original_schema: Schema { timestamp_index: 1, tsid_index: Some(0), column_schemas: ColumnSchemas { columns: [ColumnSchema { id: 1, name: "tsid", data_type: UInt64, is_nullable: false, is_tag: false, is_dictionary: false, comment: "", escaped_name: "tsid", default_value: None }, ColumnSchema { id: 2, name: "time", data_type: Timestamp, is_nullable: false, is_tag: false, is_dictionary: false, comment: "", escaped_name: "time", default_value: None }, ColumnSchema { id: 3, name: "tag1", data_type: String, is_nullable: false, is_tag: true, is_dictionary: false, comment: "", escaped_name: "tag1", default_value: None }, ColumnSchema { id: 4, name: "tag2", data_type: String, is_nullable: false, is_tag: true, is_dictionary: false, comment: "", escaped_name: "tag2", default_value: None }, ColumnSchema { id: 5, name: "value", data_type: Int8, is_nullable: false, is_tag: false, is_dictionary: false, comment: "", escaped_name: "value", default_value: None }, ColumnSchema { id: 6, name: "field2", data_type: Float, is_nullable: false, is_tag: false, is_dictionary: false, comment: "", escaped_name: "field2", default_value: None }] }, version: 1, primary_key_indexes: [0, 1] }, projection: Some([1, 2, 3, 4, 5]) }, predicate: Predicate { exprs: [BinaryExpr(BinaryExpr { left: BinaryExpr(BinaryExpr { left: Column(Column { relation: None, name: "time" }), op: Lt, right: Literal(TimestampMillisecond(1691974518000, None)) }), op: And, right: BinaryExpr(BinaryExpr { left: Column(Column { relation: None, name: "tag1" }), op: Eq, right: Literal(Utf8("test_tag")) }) })], time_range: TimeRange { inclusive_start: Timestamp(-9223372036854775808), exclusive_end: Timestamp(1691974518000) } }, metrics_collector: MetricsCollector { name: "", metrics: Mutex { data: [], poisoned: false, .. }, children: Mutex { data: [], poisoned: false, .. } } } }), (TableIdentifier { catalog: "test_catalog", schema: "test_schema", table: "__test_3" }, UnresolvedSubTableScan { table: TableIdentifier { catalog: "test_catalog", schema: "test_schema", table: "__test_3" }, read_request: ReadRequest { request_id: RequestId(42), opts: ReadOptions { batch_size: 10000, read_parallelism: 8, deadline: None }, projected_schema: ProjectedSchema { original_schema: Schema { timestamp_index: 1, tsid_index: Some(0), column_schemas: ColumnSchemas { columns: [ColumnSchema { id: 1, name: "tsid", data_type: UInt64, is_nullable: false, is_tag: false, is_dictionary: false, comment: "", escaped_name: "tsid", default_value: None }, ColumnSchema { id: 2, name: "time", data_type: Timestamp, is_nullable: false, is_tag: false, is_dictionary: false, comment: "", escaped_name: "time", default_value: None }, ColumnSchema { id: 3, name: "tag1", data_type: String, is_nullable: false, is_tag: true, is_dictionary: false, comment: "", escaped_name: "tag1", default_value: None }, ColumnSchema { id: 4, name: "tag2", data_type: String, is_nullable: false, is_tag: true, is_dictionary: false, comment: "", escaped_name: "tag2", default_value: None }, ColumnSchema { id: 5, name: "value", data_type: Int8, is_nullable: false, is_tag: false, is_dictionary: false, comment: "", escaped_name: "value", default_value: None }, ColumnSchema { id: 6, name: "field2", data_type: Float, is_nullable: false, is_tag: false, is_dictionary: false, comment: "", escaped_name: "field2", default_value: None }] }, version: 1, primary_key_indexes: [0, 1] }, projection: Some([1, 2, 3, 4, 5]) }, predicate: Predicate { exprs: [BinaryExpr(BinaryExpr { left: BinaryExpr(BinaryExpr { left: Column(Column { relation: None, name: "time" }), op: Lt, right: Literal(TimestampMillisecond(1691974518000, None)) }), op: And, right: BinaryExpr(BinaryExpr { left: Column(Column { relation: None, name: "tag1" }), op: Eq, right: Literal(Utf8("test_tag")) }) })], time_range: TimeRange { inclusive_start: Timestamp(-9223372036854775808), exclusive_end: Timestamp(1691974518000) } }, metrics_collector: MetricsCollector { name: "", metrics: Mutex { data: [], poisoned: false, .. }, children: Mutex { data: [], poisoned: false, .. } } } })], partition_count=3

